### PR TITLE
extend warning if globs are used instead of regex to local hooks

### LINKED ISSUE #2521: warnings for regexes don't apply to `repo: local`
looks like these were left out -- should be a pretty easy patch similar to #1707 / #1750

### PR DESCRIPTION
**Pull request by chrisRedwine**
Resolves #2521 
Apologies for the diff - those two classes were just moved up w/o modification

### CODE REVIEW

File: pre_commit/clientlib.py [hunk 1]
```diff
@@ -81,6 +118,8 @@ def _make_argparser(filenames_help: str) -> argparse.ArgumentParser:
     cfgv.Optional('require_serial', cfgv.check_bool, False),
     cfgv.Optional('stages', cfgv.check_array(cfgv.check_one_of(C.STAGES)), []),
     cfgv.Optional('verbose', cfgv.check_bool, False),
+    OptionalSensibleRegexAtHook('files', cfgv.check_string),
+    OptionalSensibleRegexAtHook('exclude', cfgv.check_string),
```
Comment (asottile) [inline pre_commit/clientlib.py:122]: we don't want to warn users about things outside their control (remote repositories doing this incorrectly)

Comment (chrisRedwine) [reply pre_commit/clientlib.py:122]: makes sense, thanks for the direction. I think I have a better understanding of the `cfgv` stuff after looking at it a bit, so I'll post a revision here in a minute.

Comment (chrisRedwine) [reply pre_commit/clientlib.py]: I reworked it per your call-out. 

Also noticed that the existing validators for this were calling `super().check` and passing a check func of `check_string`, which as far as I can tell is superfluous work since there are already validators on `CONFIG_HOOK_DICT` and `CONFIG_SCHEMA` for `files` & `exclude` which use `check_string_regex`. So, I removed the `super().check` calls and just passed `check_any` instead, similar to how you did the `NotAllowed` validator. Let me know if that doesn't jive, though. 

Thanks!

Comment (asottile) [reply pre_commit/clientlib.py]: compare:

```python
x = OptionalSensibleRegexAtHook('files', cfgv.check_string)
x.check({'files': None})
```

with super:

```console
$ python3 t.py 
Traceback (most recent call last):
  File "/tmp/add-trailing-comma/t.py", line 25, in <module>
    x.check({'files': None})
  File "/tmp/add-trailing-comma/t.py", line 7, in check
    super().check(dct)
  File "/tmp/add-trailing-comma/venv/lib/python3.10/site-packages/cfgv.py", line 53, in _check_optional
    with validate_context(f'At key: {self.key}'):
  File "/usr/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/add-trailing-comma/venv/lib/python3.10/site-packages/cfgv.py", line 34, in validate_context
    raise ValidationError(e, ctx=msg).with_traceback(tb) from None
  File "/tmp/add-trailing-comma/venv/lib/python3.10/site-packages/cfgv.py", line 31, in validate_context
    yield
  File "/tmp/add-trailing-comma/venv/lib/python3.10/site-packages/cfgv.py", line 54, in _check_optional
    self.check_fn(dct[self.key])
  File "/tmp/add-trailing-comma/venv/lib/python3.10/site-packages/cfgv.py", line 325, in check_type_fn
    raise ValidationError(
cfgv.ValidationError: 
==> At key: files
=====> Expected string got NoneType
```

without super:

```console
$ python3 t.py 
Traceback (most recent call last):
  File "/tmp/add-trailing-comma/t.py", line 25, in <module>
    x.check({'files': None})
  File "/tmp/add-trailing-comma/t.py", line 9, in check
    if '/*' in dct.get(self.key, ''):
TypeError: argument of type 'NoneType' is not iterable
```

Comment (chrisRedwine) [reply pre_commit/clientlib.py]: roger that - made the update

Comment (chrisRedwine) [reply pre_commit/clientlib.py]: I added a test for the counterexample. Definitely need to make sure I can commit the same amount of attention I would to closed source in poking around / exploring edge cases / understanding the code base / writing tests / etc., rather than letting the novelty & excitement of open source tempt me to just grab a ticket and push something quickly. So, I apologize & appreciate the patience here - I'll do better!

**Review by asottile — APPROVED** [review]
